File: sdk/js/src/verificationReport.ts
/**
 * What verification found — INCLUDING what it did not look at.
 *
 * `notChecked` is a field, not a footnote, and this binding passes it through
 * untouched. A "valid" verdict that quietly skipped revocation is worse than no
 * verifier at all: it turns a missing capability into a false assurance, which
 * is exactly the trust a signing feature sells. Dropping it on the way through
 * an SDK would be the same lie one layer up.
 *
 * The four checks stay separate for the same reason. "The signature is valid
 * but covers only part of the file" is a different fact from "the signature is
 * wrong", and a caller that cannot tell them apart cannot explain the answer to
 * anyone.
 */

/** The outcome of one check: passed, or failed with the reason. */
export class Check {
  readonly status: string | null;
  readonly reason: string | null;

  constructor(item: unknown) {
    const payload = asRecord(item);
    this.status = typeof payload.status === 'string' ? payload.status : null;
    this.reason = typeof payload.reason === 'string' ? payload.reason : null;
  }

  /** A PARTICIPLE stands alone — the predicate rule the six mirror. */
  get passed(): boolean {
    return this.status === 'passed';
  }

  toString(): string {
    return this.reason ? `${this.status}: ${this.reason}` : String(this.status);
  }
}

/** The four checks, the verdict, and the list of what was never looked at. */
export class VerificationReport {
  readonly signature: Check;
  readonly coverage: Check;
  readonly certificateValidity: Check;
  readonly trustChain: Check;
  readonly notChecked: readonly string[];
  private readonly validFlag: unknown;

  constructor(payload: Record<string, unknown>) {
    this.validFlag = payload.valid;
    this.signature = new Check(payload.signature);
    this.coverage = new Check(payload.coverage);
    this.certificateValidity = new Check(payload.certificateValidity);
    this.trustChain = new Check(payload.trustChain);
    this.notChecked = Array.isArray(payload.notChecked)
      ? payload.notChecked.map((entry) => String(entry))
      : [];
  }

  static parse(payload: string): VerificationReport {
    return new VerificationReport(asRecord(JSON.parse(payload)));
  }

  /**
   * Whether every check this release PERFORMS passed.
   *
   * Read `notChecked` beside it: this is not "the document is trustworthy", it
   * is "nothing we looked at was wrong".
   */
  get valid(): boolean {
    return this.validFlag === true;
  }

  /** The four checks by name. A Map, so a wire key can never reach a prototype. */
  get checks(): Map<string, Check> {
    return new Map([
      ['signature', this.signature],
      ['coverage', this.coverage],
      ['certificateValidity', this.certificateValidity],
      ['trustChain', this.trustChain],
    ]);
  }
}

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
}

File: sdk/python/src/shojiku/verification_report.py
"""What verification found — INCLUDING what it did not look at.

``not_checked`` is a field, not a footnote, and this binding passes it through
untouched. A "valid" verdict that quietly skipped revocation is worse than no
verifier at all: it turns a missing capability into a false assurance, which is
exactly the trust a signing feature sells. Dropping it on the way through an SDK
would be the same lie one layer up.

The four checks stay separate for the same reason. "The signature is valid but
covers only part of the file" is a different fact from "the signature is wrong",
and a caller that cannot tell them apart cannot explain the answer to anyone.
"""

from __future__ import annotations

import json
from typing import Any


class Check:
    """The outcome of one check: passed, or failed with the reason."""

    def __init__(self, item: dict[str, Any] | None) -> None:
        payload = item or {}
        self.status: str | None = payload.get("status")
        self.reason: str | None = payload.get("reason")

    @property
    def passed(self) -> bool:
        return self.status == "passed"

    def __str__(self) -> str:
        return f"{self.status}: {self.reason}" if self.reason else str(self.status)


class VerificationReport:
    """The four checks, the verdict, and the list of what was never looked at."""

    def __init__(self, payload: dict[str, Any]) -> None:
        self._valid = payload.get("valid")
        self.signature = Check(payload.get("signature"))
        self.coverage = Check(payload.get("coverage"))
        self.certificate_validity = Check(payload.get("certificateValidity"))
        self.trust_chain = Check(payload.get("trustChain"))
        self.not_checked: tuple[str, ...] = tuple(payload.get("notChecked") or ())

    @staticmethod
    def parse(payload: str) -> VerificationReport:
        return VerificationReport(json.loads(payload))

    @property
    def valid(self) -> bool:
        """Whether every check this release PERFORMS passed.

        Read ``not_checked`` beside it: this is not "the document is
        trustworthy", it is "nothing we looked at was wrong".
        """
        return self._valid is True

    @property
    def checks(self) -> dict[str, Check]:
        return {
            "signature": self.signature,
            "coverage": self.coverage,
            "certificate_validity": self.certificate_validity,
            "trust_chain": self.trust_chain,
        }
